File: problems/Arrays and Hashing/383: Ransom Note.js
// https://leetcode.com/problems/ransom-note/description/
// Difficulty: Easy

// Solution
// O(n) time and O(1) space. Create a mapping of all the characters in the magazine, then iterate over the ransom note and decrement the mapping. If the count in the mapping is ever negative, return false.

var canConstruct = function (ransomNote, magazine) {
  const chars = "abcdefghijklmnopqrstuvwxyz";
  const mapping = {};
  for (const char of chars) {
    mapping[char] = 0;
  }
  for (const char of magazine) {
    mapping[char]++;
  }

  for (const char of ransomNote) {
    mapping[char]--;
    if (mapping[char] < 0) {
      return false;
    }
  }

  return true;
};

File: problems/Leetcode/383: Ransom Note.py
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        ransomhash = collections.defaultdict(int)
        maghash = collections.defaultdict(int)

        for char in ransomNote:
            ransomhash[char] += 1
        
        for char in magazine:
            maghash[char] += 1
            
        return ransomhash.items() <= maghash.items()